rosbag2_py/src/rosbag2_py/_transport.cpp

```diff
@@ -190,23 +190,39 @@ class Player
 class Recorder
 {
 private:
-  std::unique_ptr<rclcpp::executors::SingleThreadedExecutor> exec_;
+  using ExecutorT = rclcpp::executors::SingleThreadedExecutor;
+  static std::unique_ptr<ExecutorT, void (*)(ExecutorT *)> & get_executor_instance()
+  {
+    static std::unique_ptr<ExecutorT, void (*)(ExecutorT *)> executor(
+      new ExecutorT(),
+      [](ExecutorT * exec_ptr) {
+        delete exec_ptr;
+        rclcpp::shutdown();
+      }
+    );
+    return executor;
+  }
+  std::shared_ptr<rosbag2_transport::Recorder> recorder_;
 
 public:
   Recorder()
   {
-    rclcpp::init(0, nullptr);
-    exec_ = std::make_unique<rclcpp::executors::SingleThreadedExecutor>();
+    auto init_options = rclcpp::InitOptions();
+    init_options.shutdown_on_signal = false;
+    if (!rclcpp::ok()) {
+      rclcpp::init(0, nullptr, init_options, rclcpp::SignalHandlerOptions::None);
+    }
     std::signal(
       SIGTERM, [](int /* signal */) {
-        rclcpp::shutdown();
+        rosbag2_py::Recorder::cancel();
+      });
```

Comment:
I guess I still don't understand why this has to be static, singleton globals. Why not e.g.

```
std::signal(
  SIGTERM, [this](int) {
    this->cancel();
  });
```

Or, does the instance get destroyed before the signal handler is called?

Reply:
Well, just because 
```
std::signal(
  SIGTERM, [this](int) {
    this->cancel();
  });
```
will not compile.
We can't pass `this` pointer in this case. `std::signal(..)`  expecting a pointer to the static function.
i.e. `this->cancel();` shall be static.

After making a deeper analysis and hefty debug session of the failure on CI referred in the upper comment https://github.com/ros2/rosbag2/pull/1301#issuecomment-1577729842 I found out that there sometimes appear deadlock inside `SingleThreadedExecutor` when calling `executor->cancel()` method from another thread.

I will create a follow-up issue for the executor, however for a while considering to rewrite this PR without a static instance of the executor. As a workaround, I am considering triggering my static `exit_` variable inside signal handlers and just waiting in the main thread when the executor will finish its `spin_all(std::chrono::milliseconds(10));` operation.